### ISSUE TITLE
Underlining or bolding specific parts of text that matches in results.

### ISSUE THREAD

**Issue by jasonxeno** [miguelalarcos/xautocomplete#2]
So Arunoda from Meteorhacks has this package called search-source:

https://github.com/meteorhacks/search-source
http://instant-search-demo.meteor.com/
https://meteorhacks.com/implementing-an-instant-search-solution-with-meteor.html

sample app code using the package:
https://github.com/meteorhacks-samples/meteor-instant-search-demo

It's similar to this package with searching but it specializes in search results on a page, not auto-completing forms with autoform support like yours.

One cool thing about it, is that it **bolds** parts of the string that the results match.

So if you type 'Den' in the author box, the results would show **Den**net, Daniel

Field: Denn
Results: **Denn**et, Daniel

Field: Denne
Results: **Denne**t, Daniel

Field: Dennet
Results: **Dennet**, Daniel

etc...

This package does a similar thing but underlines the part of the string that matches:
https://github.com/comerc/meteor-autoform-placecomplete

![demo](https://cloud.githubusercontent.com/assets/6493925/5790440/2f7afe74-9e5e-11e4-9f73-1c6df7d1d498.gif)

I posted an issue in the search-source package about form support for the search-source package here:
https://github.com/meteorhacks/search-source/issues/5

So maybe he can give us some advice.

Do you have any ideas on how to do this in your package?

Looking in the search-source code:
https://github.com/meteorhacks-samples/meteor-instant-search-demo/blob/master/client/app.js

I think this is the part where the matchText gets bolded:

``` javascript
Template.searchResult.helpers({
  getPackages: function() {
    return PackageSearch.getData({
      transform: function(matchText, regExp) {
        return matchText.replace(regExp, "<b>$&</b>")
      },
      sort: {isoScore: -1}
    });
  },
```

Any ideas on how to replicate bolding the match text in your xautocomplete package?

Thank you!

**Comment by miguelalarcos** [comment]
I have been thinking in this feature, and I think the only way is in renderfunction, creating the template html directly. I hope to do this soon. Thanks for your info.

**Comment by jasonxeno** [comment]
Awesome! Yeah I think there could be something in renderfunction that can call a transform function and then put the meta data about that in the data context of the xautocomplete object, which can then render in the `<b>'s or <u>'s` in the output of the string values of the doc. I'm still looking through the search-source code and the search-source example app. I believe this is the actual transform function that is called:

line 120:
https://github.com/meteorhacks/search-source/blob/de893d5598235ebc90bedd6a9dc2d00ee48abcf0/lib/client.js

``` javascript
  function transform(doc) {
    self.searchFields.forEach(function(field) {
      if(self.currentQuery && doc[field]) {
        if(options.transform) {
          doc[field] = options.transform(doc[field], regExp, field, self.currentQuery);
        }
      }
    });
    return doc;
  }
```

Coffeescript conversion:

``` coffeescript
transform = (doc) ->
  self.searchFields.forEach (field) ->
    doc[field] = options.transform(doc[field], regExp, field, self.currentQuery)  if options.transform  if self.currentQuery and doc[field]
    return

  doc
```

And then this is an example of how to call the transform function:

https://github.com/meteorhacks-samples/meteor-instant-search-demo/blob/530827ec122e7d678075410f1de1f0b397c77881/client/app.js

``` javascript
Template.searchResult.helpers({
  getPackages: function() {
    return PackageSearch.getData({
      transform: function(matchText, regExp) {
        return matchText.replace(regExp, "<b>$&</b>")
      },
      sort: {isoScore: -1}
    });
  },
});
```

Coffeescript conversion:

``` coffeescript
Template.searchResult.helpers
  getPackages: ->
    PackageSearch.getData
      transform: (matchText, regExp) ->
        matchText.replace regExp, "<b>$&</b>"

      sort:
        isoScore: -1
```

**Comment by miguelalarcos** [comment]
this code does the trick:

``` coffee
@renderAuthors = (x, query) ->
  txt = '<td>' +x.name + ' ' + x.surname+ '</td>'
  txt.replace(query, "<b>$&</b>")
```

I have to upload to atmosphere the new version of xautocomple. I upload right now to git.

**Comment by jasonxeno** [comment]
Oh neat thanks! I'll test it out now and check out your commit.

**Comment by miguelalarcos** [comment]
uploaded to atmosphere. v0.1.2

**Comment by jasonxeno** [comment]
I tested and the new feature works great! Very excited about this, thanks so much. I'll keep testing the package for you.

**Comment by miguelalarcos** [comment]
Thanks for your help.

**Comment by jasonxeno** [comment]
Just discovered a way to make the bold feature case insensitive. Make a new RegEx variable with option 'gi'. Then it will work regardless of the case of the letters.

``` coffeescript
@renderCity = (x, query) ->
  txt = "<td>#{x.city}, #{x.state} #{x.zip}</td>"
  pattern = new RegExp(query, 'gi');
  txt.replace(pattern, "<b>$&</b>")
```

**Comment by miguelalarcos** [comment]
Great !!

**Comment by miguelalarcos** [comment]
By the way, what about this?

``` coffee
@renderCity = (x, query) ->
  txt = "<td>#{x.city}</td><td>#{x.state}</td><td>#{x.zip}</td>"
  pattern = new RegExp(query, 'gi');
  txt.replace(pattern, "<b>$&</b>")
```

How do you see it?

**Comment by jasonxeno** [comment]
Aha! Thank you @miguelalarcos, now it looks much nicer because everything is aligned.

![screenshot3](https://cloud.githubusercontent.com/assets/6493925/5791346/ea5751f2-9e95-11e4-8040-27ecf86d56ad.png)

**Comment by miguelalarcos** [comment]
That looks wonderful!!! Are you using semantic-ui? (The package is intended to be used with semantic-ui)

Now there is a new version v0.1.3 where I have included two new keys: renderkey and valuekey to be used instead of renderfunction and valuefunction if just the property of the object is enough.

``` html
   Simple autocomplete: {{> xautocomplete formid='2' name='surname' renderkey='surname' call='authors' callbackfunction='myCallback'}}
    <br>
        Reference: {{> xautocomplete   reference='authors' formid='3' name='authorId' valuekey='surname' renderfunction='renderAuthors' call='authors'}}
    <br>
```

(Please not that valuefunction or valuekey is only needed when reference key is set)

**Comment by elGusto** [comment]
@jasonxeno How do you define your data source to search in your collection ?

I am trying to use the xautocomplete call attribute by defining a meteor method but I am stuck at this point. How do you call the search-source api from there ?

**Comment by miguelalarcos** [comment]
@elGusto let me explain how to do it. Given that you have the attribute _call_ with the value _authors_:

``` html
{{> xautocomplete formid='2' name='surname' renderkey='surname' call='authors' callbackfunction='myCallback'}}
```

 then you need a method like this in the server:

``` coffee
Meteor.methods
  authors: (query)->
    if query != ''
      authors.find(surname: {$regex: '^.*'+query+'.*$'}).fetch()
    else
      []
```

Let me know any doubts.

**Comment by jasonxeno** [comment]
@miguelalarcos thank you!! I'm actually using bootstrap and then this on top of bootstrap:
http://www.keenthemes.com/preview/metronic/theme/templates/admin2/

I started the project before Semantic UI got more advanced so that is why I'm using bootstrap, it looks like it is getting better and better!

Here is my current css to get that style in bootstrap:

``` css
.xpopover {
    position: absolute;
    background-color: #ffffff;
    z-index: 1;
    border: 1px solid #ccc;
    border-top-color: #d9d9d9;
    box-shadow: 0 2px 4px rgba(0,0,0,0.2);
}

.xitem > td {
    padding: 0 5px !important;;
}

.xwidget {
    height: 22px !important;
}

.xselected{
    background-color: #ffff00;
    font-weight: 600;
}

.xclose {
    cursor: pointer;
}
```

and I'm also changing the classes like so with package aldeed:template-extension:

``` coffeescript
Template.onRendered ->
  @$(".ui").removeClass "ui table"
  @$(".xautocomplete-input").addClass "form-control"
```

**Comment by miguelalarcos** [comment]
@jasonxeno ,  great!!!

**Comment by miguelalarcos** [comment]
@jasonxeno I am adding a new feature to the package and I would like to know your opinion.

I don't like to define globals vars, so I have thought to use just one, _settings_ and use this way:

``` html
Reference: {{> xautocomplete formid='3' name='authorId' settings='settings2'}}
```

and in your code:

``` coffee
@settings2 =
  renderFunction: renderAuthors
  reference : 'authors'
  valueKey: 'surname'
  call : 'authors'
```

Thanks for your attention.

**Comment by jasonxeno** [comment]
@miguelalarcos That new way looks great! Wonderful idea. The global vars get a bit messy, and that way is cleaner. Let me know when you got the new way working and I'll test it out!!

**Comment by miguelalarcos** [comment]
I have uploaded to atmosphere the new version. All tests have passed but there could be bugs I haven't noticed.

**Comment by miguelalarcos** [comment]
@jasonxeno I have released a new version. The last one had several bugs. I have tested with new tests form the mode settings. Thanks for your attention.

**Comment by jasonxeno** [comment]
@miguelalarcos looks like you have been busy coding! I'm planning on testing the latest version this afternoon for you.

**Comment by miguelalarcos** [comment]
Thanks a lot for your help!  :)

**Comment by jasonxeno** [comment]
@miguelalarcos I converted all my xautocomplete fields to the new settings feature and it works great, I haven't found any issues. Great work!

**Comment by miguelalarcos** [comment]
Thank you for contributing to the project!

**Comment by miguelalarcos** [comment]
@jasonxeno  I would like to know your opinion about a feature of the xautocomplete package. It is when you use it in simple mode (no reference, no multiple). Should it return any thing you type in the input or should it return null when de text is not in the items returned by the remote call? What do you think is better? Thanks :)

**Comment by miguelalarcos** [comment]
@jasonxeno , I have released a new version of xautocomplete (0.3.0) and it breaks compatibility backwards.
Now, you have to use true or false instead of 'true' or 'false' in the templates.
And there's no more valueFunctioin or valueKey. Instead there is a fieldref attribute.

**Comment by jasonxeno** [comment]
@miguelalarcos Thanks for the heads up! Good changes. xautocomplete has been working well for me!

Looks like I missed the message above. My opinion is usually more explicit info is better, for example return an object with both points of information. Then the behavior of each field in the object is consistent.

object =
   input_text: 'text in the input'
   items_returned: true

and

object =
   input_text: 'text in the input'
   items_returned: false
